### PR TITLE
Allow up to maxRedirects upon receiving HTTP 301 status

### PR DESCRIPTION
**Pull request by k0ral**
The [`roundTripWithOptionalFollowRedirect()` method](https://github.com/google/go-github/blob/b45ef89f2dd0ec803ab8f27725bdb4941c0cbc99/github/github.go#L1561) currently takes a boolean parameter `followRedirects`, used to support HTTP 301 redirections. However, at most 1 redirection is handled at the moment (the parameter is set to `false` after the first iteration.)

I've stumbled upon multiple repositories where `https://api.github.com/repos/<OWNER>/<PROJECT>/branches/<BRANCH>` redirects twice ; as a result, the [`GetBranch()` method](https://github.com/google/go-github/blob/b45ef89f2dd0ec803ab8f27725bdb4941c0cbc99/github/repos.go#L1277) only follows the first redirection, then returns the 2nd HTTP 301 response.

This MR proposes to generalize the `followRedirects` boolean into a `maxRedirects` integer, to support an arbitrary number of redirections. This impacts multiple methods, not just `GetBranch()`.

:warning: This is a breaking change.

### CODE REVIEW

**Review by gmlewis — APPROVED** [review]
Thank you, @k0ral !
LGTM.

Awaiting second LGTM+Approval from any other contributor to this repo before merging.

**Review by valbeat — APPROVED** [review]

**Comment by gmlewis** [comment]
Thank you, @valbeat !
Merging.

**Comment by WillAbides** [comment]
I know this is already merged, but I would like to suggest that since we are doing a breaking change for these methods we consider updating them to accept an options struct instead of just the maxRedirects option. This will allow future options to be added without a breaking change. If they previously used an options struct this wouldn't need to be a breaking change. We can't avoid a break now, but we can avoid future breaks.

**Comment by gmlewis** [comment]
Good idea, @WillAbides - do you want to make a PR that addresses this and we will hold off on the next release until it is merged?

**Comment by gmlewis** [comment]
@WillAbides - I was [requested](https://github.com/google/go-github/pull/2947#issuecomment-1756982704) to cut a new release, but just remembered that [you wanted](https://github.com/google/go-github/pull/2939#issuecomment-1747392285) to change this `maxRedirects` API to use options instead, and thereby break the API twice before a new release instead of breaking the API twice between two different major releases.

Do you want me to hold off on cutting the new release, or shall we simply go ahead and break it twice?

I'm leaning toward breaking it twice at the moment, since I like to be responsive when users want a new release and it has been about a month since the last one.

**Comment by gmlewis** [comment]
I'll go ahead on the release. We can work on an improved interface for the next breaking release.

**Comment by WillAbides** [comment]
@gmlewis I agree with going ahead with the release because the need to move to options is hypothetical at this point.

**Pull request merged**
